File: sim_engine/src/rag/prompts.rs
//! Prompt templates for the counselling advisor.

/// System prompt injected at the start of every LLM conversation.
pub fn system_prompt() -> &'static str {
    "You are a helpful JoSAA counselling advisor for students appearing in JEE \
     (Joint Entrance Examination) in India. You assist with:\n\
     - The JoSAA counselling and seat allotment process\n\
     - Document requirements for reporting to institutes\n\
     - Choice filling strategy and deadlines\n\
     - Quota rules (AI, HS, EWS, OBC-NCL, SC, ST, PwD)\n\
     - Withdrawal, upgradation, and freeze/float/slide options\n\
     - Reading and interpreting their uploaded counselling documents and screenshots\n\
     \n\
     When answering:\n\
     - Prioritise information from the uploaded context documents\n\
     - If a screenshot of a counselling portal is attached, reference the specific \
       details visible in it\n\
     - If the context is insufficient, clearly say so and supplement with general knowledge\n\
     - Be concise, accurate, and encouraging\n\
     - Format steps and lists clearly\n\
     - Always advise students to verify critical details on the official JoSAA portal \
       (josaa.nic.in)"
}

/// Build the user-facing message containing retrieved context and the question.
pub fn user_prompt(context_chunks: &[String], question: &str) -> String {
    if context_chunks.is_empty() {
        return format!(
            "Question: {question}\n\n\
             Note: No documents have been uploaded yet. \
             I will answer from general knowledge."
        );
    }

    let context = context_chunks
        .iter()
        .enumerate()
        .map(|(i, chunk)| format!("[Excerpt {}]\n{chunk}", i + 1))
        .collect::<Vec<_>>()
        .join("\n\n---\n\n");

    format!(
        "Relevant excerpts from uploaded documents:\n\n\
         {context}\n\n\
         ════════════════════════════════\n\n\
         Question: {question}\n\n\
         Answer using the excerpts above where possible. \
         Note which excerpt supports each point."
    )
}

File: rag_service/app/rag/prompts.py
"""System and user prompt templates for the JoSAA counselling advisor."""


SYSTEM_PROMPT = """\
You are a helpful JoSAA counselling advisor for students appearing in JEE \
(Joint Entrance Examination) in India. You assist with:
- The JoSAA counselling and seat allotment process
- Document requirements for reporting to institutes
- Choice filling strategy and deadlines
- Quota rules (AI, HS, EWS, OBC-NCL, SC, ST, PwD)
- Withdrawal, upgradation, and freeze/float/slide options
- Reading and interpreting their uploaded counselling documents and screenshots

When answering:
- Prioritise information from the uploaded context documents
- If a screenshot of a counselling portal is attached, reference the specific \
details visible in it
- If the context is insufficient, clearly say so and supplement with general knowledge
- Be concise, accurate, and encouraging
- Format steps and lists clearly
- Always advise students to verify critical details on the official JoSAA portal \
(josaa.nic.in)"""


def build_user_prompt(context_chunks: list[str], question: str) -> str:
    """
    Build the user-facing message that includes retrieved context excerpts
    and the user's question.
    """
    if not context_chunks:
        return (
            f"Question: {question}\n\n"
            "Note: No documents have been uploaded yet. "
            "I will answer from general knowledge."
        )

    context = "\n\n---\n\n".join(
        f"[Excerpt {i + 1}]\n{chunk}" for i, chunk in enumerate(context_chunks)
    )
    return (
        f"Relevant excerpts from uploaded documents:\n\n"
        f"{context}\n\n"
        f"════════════════════════════════\n\n"
        f"Question: {question}\n\n"
        f"Answer using the excerpts above where possible. "
        f"Note which excerpt supports each point."
    )
